### PR TITLE
Modular Assistive Driving System (MADS)

### PR DESCRIPTION
**Pull request by devtekve**
Refactor MADS safety with improved state management

# Supported makes
- [x] Hyundai (_+ LKAS button_)
- [x] Toyota (_No LKAS button_)
- [x] Honda
- [x] Ford
- [x] Chrysler
- [x] GM
- [x] Mazda
- [x] Nissan
- [x] Subaru
- [x] Volkswagen
~- [ ] Tesla~

### Tested real world 
- [x] Hyundai (_+ LKAS button_)
- [x] Toyota (_No LKAS button_)
- [ ] Honda
- [x] Ford
- [x] Chrysler
- [ ] GM
- [ ] Mazda
- [ ] Nissan
- [ ] Subaru
- [ ] Volkswagen
~- [ ] Tesla~

_**Note: Real world tests will be deferred to the SP PR**_

# Changes
## Core Changes:
- Introduced a MADSState struct to centralize state management
- Removed global state variables in favor of structured state
- Implemented button transition handling with explicit state tracking (PRESSED/RELEASED/NO_CHANGE)
- Added state flags for button availability detection
- Simplified lateral control permission logic

## Button Handling:
- Separated main button and LKAS button state tracking
- Added independent engagement states for each button
- Improved button press detection across multiple platforms
- Added support for main and LKAS buttons on Hyundai platforms
- Modified ACC main state handling

## Testing:
- Added comprehensive test coverage for MADS state transitions
- Added new MADS-specific test base class for consistent testing across platforms
- Added mutation testing for state management
- Extended timeout for mutation tests from 5 to 8 minutes
- Added extensive button press validation tests
- Enhanced debugging output in replay drive tests

The refactored code provides a more organized implementation of MADS safety features while maintaining compatibility with existing safety checks.

# Relates to: 
* SP PR: https://github.com/sunnypilot/sunnypilot/pull/446
* Opendbc PR: https://github.com/sunnypilot/opendbc/pull/29

### CODE REVIEW

**Comment by devtekve** [comment]
Tested a ford route and I see a lot of messages blocked, we need to validate the code flow, might be missing something. Tested `F150Lightning`

**Comment by devtekve** [comment]
I should probably find a way to make the tests fail if the make isn't working as expected. Right now the tests mainly focus on ensuring that when MADS is not enabled, everything behaves like stock OP. But for example, I tested a ford route when I was con commit https://github.com/sunnypilot/panda/pull/40/commits/86f52054a3900bffd7d579dfc9e1462dd196ac54 and while [all the tests passed](https://github.com/sunnypilot/panda/actions/runs/12135740390) I manually ran a `replay_drive` and found several messages blocked. Which hints me that steering messages were blocked even though mads was active and it's slipping through the safety tests. (Again, not a safety concern, because the goal is to block by default if something is not right, so likely something is not right on the ford code yet)

**Comment by devtekve** [comment]
From what I can see, toyota seems to be good (no LKAS button for now)

**Comment by sunnyhaibin** [comment]
Working on Toyota, Subaru, Honda as final

File: board/safety/safety_subaru.h [hunk 1]
```diff
@@ -106,10 +106,19 @@ static void subaru_rx_hook(const CANPacket_t *to_push) {
     update_sample(&angle_meas, angle_meas_new);
   }
 
+  if ((addr == MSG_SUBARU_ES_LKAS_State) && (bus == SUBARU_CAM_BUS)) {
+    bool lkas_hud = ((GET_BYTE(to_push, 2) & 0x0CU) >> 2U) > 0U;
+    if (lkas_hud) {
+      lkas_button_press = MADS_BUTTON_PRESSED;
+    }
+  }
+
```
Comment (sunnyhaibin) [inline board/safety/safety_subaru.h:115]:
```suggestion
  if ((addr == MSG_SUBARU_ES_LKAS_State) && (bus == SUBARU_CAM_BUS)) {
    int lkas_hud = (GET_BYTE(to_push, 2U) & 0x0CU) >> 2U;
    if (lkas_hud > 0) {
      lkas_button_press = MADS_BUTTON_PRESSED;
    }
  }

```

File: board/safety/safety_toyota.h [hunk 2]
```diff
@@ -64,6 +64,17 @@ static bool toyota_get_quality_flag_valid(const CANPacket_t *to_push) {
...
+    int addr = GET_ADDR(to_push);
+  
+    if (addr == 0x412) {
+      bool lkas_hud_visible = (GET_BYTE(to_push, 0U) & 0xC0U) > 0U;
+      if (lkas_hud_visible) {
+        lkas_button_press = MADS_BUTTON_PRESSED;
+      }
+    }
+  }
+
```
Comment (sunnyhaibin) [inline board/safety/safety_toyota.h:77]: This currently does not reference 6th and 7th bit, need to right-shift 6 positions.
```suggestion
  if (GET_BUS(to_push) == 2U) {
    int addr = GET_ADDR(to_push);
  
    if (addr == 0x412) {
      int lkas_hud_visible = (GET_BYTE(to_push, 0U) & 0xC0U) >> 6U;
      if (lkas_hud_visible > 0) {
        lkas_button_press = MADS_BUTTON_PRESSED;
      }
    }
  }

```

File: tests/libpanda/safety_helpers.h [hunk 3]
```diff
@@ -35,6 +35,26 @@ bool get_controls_allowed(void){
...
+	mads_set_system_state(enable_mads, disengage_lat_on_brake);
+}
+
+bool get_enable_mads(void){
+	return _mads_state.system_enabled;
+}
+
+bool get_disengage_lat_on_brake(void){
+	return _mads_state.disengage_lateral_on_brake;
+}
```
Comment (sunnyhaibin) [inline tests/libpanda/safety_helpers.h:56]:
```suggestion
bool get_lat_active(void){
  return is_lat_active();
}

bool get_controls_allowed_lat(void){
  return mads_is_lateral_control_allowed_by_mads();
}

void set_enable_mads(bool enable_mads, bool disengage_lat_on_brake){
  mads_set_system_state(enable_mads, disengage_lat_on_brake);
}

bool get_enable_mads(void){
  return _mads_state.system_enabled;
}

bool get_disengage_lat_on_brake(void){
  return _mads_state.disengage_lateral_on_brake;
}
```

File: tests/libpanda/safety_helpers.h [hunk 4]
```diff
@@ -181,6 +205,70 @@ bool get_honda_fwd_brake(void){
...
+  return _mads_state.main_button.transition;
+}
+
+int get_mads_main_button_current(void){
+    return *_mads_state.main_button.current;
+}
+
+int get_mads_main_button_last(void){
+    return _mads_state.main_button.last;
+}
```
Comment (sunnyhaibin) [inline tests/libpanda/safety_helpers.h:230]:
```suggestion
int get_mads_main_button_transition(void){
  return _mads_state.main_button.transition;
}

int get_mads_main_button_current(void){
  return *_mads_state.main_button.current;
}

int get_mads_main_button_last(void){
  return _mads_state.main_button.last;
}
```

File: tests/safety/test_hyundai.py [hunk 5]
```diff
@@ -167,6 +209,7 @@ def _user_gas_msg(self, gas):
     values = {"CR_Vcu_AccPedDep_Pos": gas}
     return self.packer.make_can_msg_panda("E_EMS11", 0, values, fix_checksum=checksum)
 
+
```
Comment (sunnyhaibin) [inline tests/safety/test_hyundai.py:212]:
```suggestion
```

File: tests/safety/test_subaru.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 import enum
 import unittest
+
```
Comment (sunnyhaibin) [inline tests/safety/test_subaru.py:4]:
```suggestion
```

File: board/safety/safety_chrysler.h [hunk 7]
```diff
@@ -78,6 +78,8 @@ static void chrysler_rx_hook(const CANPacket_t *to_push) {
   if ((bus == das_3_bus) && (addr == chrysler_addrs->DAS_3)) {
     bool cruise_engaged = GET_BIT(to_push, 21U);
     pcm_cruise_check(cruise_engaged);
+
```
Comment (sunnyhaibin) [inline board/safety/safety_chrysler.h:81]:
```suggestion
```

File: board/safety/safety_mazda.h [hunk 8]
```diff
@@ -36,6 +36,8 @@ static void mazda_rx_hook(const CANPacket_t *to_push) {
     if (addr == MAZDA_CRZ_CTRL) {
       bool cruise_engaged = GET_BYTE(to_push, 0) & 0x8U;
       pcm_cruise_check(cruise_engaged);
+
```
Comment (sunnyhaibin) [inline board/safety/safety_mazda.h:39]:
```suggestion
```

File: board/safety/safety_subaru.h [hunk 9]
```diff
@@ -106,10 +106,19 @@ static void subaru_rx_hook(const CANPacket_t *to_push) {
...
   // enter controls on rising edge of ACC, exit controls on ACC off
   if ((addr == MSG_SUBARU_CruiseControl) && (bus == alt_main_bus)) {
     bool cruise_engaged = GET_BIT(to_push, 41U);
     pcm_cruise_check(cruise_engaged);
+
```
Comment (sunnyhaibin) [inline board/safety/safety_subaru.h:120]:
```suggestion
```

File: board/safety/safety_subaru_preglobal.h [hunk 10]
```diff
@@ -36,6 +36,8 @@ static void subaru_preglobal_rx_hook(const CANPacket_t *to_push) {
     if (addr == MSG_SUBARU_PG_CruiseControl) {
       bool cruise_engaged = GET_BIT(to_push, 49U);
       pcm_cruise_check(cruise_engaged);
+
```
Comment (sunnyhaibin) [inline board/safety/safety_subaru_preglobal.h:39]:
```suggestion
```

File: tests/safety_replay/replay_drive.py [hunk 11]
```diff
@@ -68,6 +73,7 @@ def replay_drive(lr, safety_mode, param, alternative_experience, segment=False):
   print("blocked msgs:", tx_blocked)
   print("blocked with controls allowed:", tx_controls_blocked)
   print("blocked addrs:", blocked_addrs)
+  print("Mads enabled:", safety.get_enable_mads())
```
Comment (sunnyhaibin) [inline tests/safety_replay/replay_drive.py:76]:
```suggestion
  print("mads enabled:", safety.get_enable_mads())
```

**Review by sunnyhaibin — APPROVED** [review]

**Pull request merged**